### PR TITLE
Implement POSIX socket+netdb wrapper

### PR DESCRIPTION
**Pull request by TuxSH**
This pull requests implements the remaining bsd:u/s commands (except `bsdGetResourceStatistics`, `bsdRecvMmsg`, `bsdSendMmsg`), and all `sfdnsres` commands, then wraps everything up in a   POSIX-compliant interface in `socket.c`.

The end-user should only use `socket.c`.

I've included all the necessary FreeBSD headers that are usable on the Switch, after cleaning them up.

**Bugs/TODO**:

* Fix the 0xce01 "kernel slabheap depleted" issue that arises in some common situations, for example when sending under certain conditions (ftpd is affected), or ioctls that makes bsdservices call e.g. nifm (SIOCGIFCONF for example, needed to get the Switch's IP and implement `gethostid`).
* Implement recvmmsg and sendmmsg
* Test the implementation of `netdb.h`functions, I have not tested them thoroughsly.

### CODE REVIEW

File: nx/include/switch.h [hunk 1]
```diff
@@ -33,7 +33,7 @@ extern "C" {
 #include "switch/services/apm.h"
 #include "switch/services/applet.h"
 #include "switch/services/audout.h"
-#include "switch/services/bsd.h"
+//#include "switch/services/bsd.h" Use switch/runtime/devices/socket.h instead
```
Comment (plutooo) [inline nx/include/switch.h]: Why commented?

File: nx/include/switch.h [hunk 2]
```diff
@@ -33,7 +33,7 @@ extern "C" {
...
 #include "switch/services/fatal.h"
 #include "switch/services/usb.h"
 #include "switch/services/hid.h"
```
Comment (plutooo) [inline nx/include/switch.h]: Missing sfdnsres.h

File: nx/include/switch/services/bsd.h [hunk 3]
```diff
@@ -9,6 +9,10 @@
 #include "../types.h"
 #include "../kernel/tmem.h"
 
+#include <sys/socket.h> // for socklen_t
```
Comment (plutooo) [inline nx/include/switch/services/bsd.h]: <>-includes on top of file

File: nx/source/runtime/devices/socket.c [hunk 4]
```diff
@@ -0,0 +1,1555 @@
...
+
+Result socketGetLastBsdResult(void) {
+    return g_bsdResult;
+}
+
+Result socketGetLastSfdnsresResult(void) {
+    return g_sfdnsresResult;
+}
+
+/********************************************* BSD:U FUNCTIONS AND RELATED *********************************************/
```
Comment (plutooo) [inline nx/source/runtime/devices/socket.c]: Weird comment

File: nx/source/services/bsd.c [hunk 5]
```diff
@@ -7,16 +10,26 @@
 #include "kernel/shmem.h"
 #include "kernel/rwlock.h"
 
-#define EPIPE 32
+// Complete definition of struct timeout:
+#include <sys/time.h>
```
Comment (plutooo) [inline nx/source/services/bsd.c]: <>-includes at the top

File: nx/source/services/bsd.c [hunk 6]
```diff
@@ -7,16 +10,26 @@
...
+
+#include <fcntl.h>
```
Comment (plutooo) [inline nx/source/services/bsd.c]: <>-includes at the top

File: nx/source/services/sfdnsres.c [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+    rc = resp->result;
+    return rc;
+}
+
+Result sfdnsresGetHostByName(SfdnsresRequestResults *ret, const SfdnsresConfig *config, void *out_he_serialized, const char *name) {
+    IpcCommand c;
+    ipcInitialize(&c);
+    ipcAddSendBuffer(&c, name, name == NULL ? 0 : strlen(name) + 1, 0);
+
```
Comment (plutooo) [inline nx/source/services/sfdnsres.c]: One newline too much? :P

File: nx/include/switch.h (hunk 1 above)
Comment (fincs) [reply nx/include/switch.h]: It could be said that this is part of avoiding importing ugly BSD headers by default :p

Comment (plutooo) [reply nx/include/switch.h]: Fix should be to remove ugly-bsd-header-dependency on bsd-service shim

File: nx/include/net/bpf.h [hunk 8]
```diff
@@ -0,0 +1,484 @@
+// TuxSH: removed definitions under _KERNEL ifdef blocks, modify the prototype of some functions, some other cleanup, etc.
```
Comment (plutooo) [inline nx/include/net/bpf.h]: Is there any reason to pull in this thing when literally nobody is going to use it?

File: nx/include/net/dlt.h [hunk 9]
```diff
@@ -0,0 +1,1338 @@
+/*-
```
Comment (plutooo) [inline nx/include/net/dlt.h]: Is there any reason to pull in this thing when literally nobody is going to use it?

File: nx/include/net/if.h [hunk 10]
```diff
@@ -0,0 +1,547 @@
+// Removed kernel stuff, use anon unions
```
Comment (plutooo) [inline nx/include/net/if.h]: Is this actually required by any implemented API?

File: nx/include/net/if_dl.h [hunk 11]
```diff
@@ -0,0 +1,91 @@
+/*-
```
Comment (plutooo) [inline nx/include/net/if_dl.h]: What even is this

File: nx/include/net/if_media.h [hunk 12]
```diff
@@ -0,0 +1,794 @@
+/*	$NetBSD: if_media.h,v 1.3 1997/03/26 01:19:27 thorpej Exp $	*/
```
Comment (plutooo) [inline nx/include/net/if_media.h]: What even is this

File: nx/include/net/route.h [hunk 13]
```diff
@@ -0,0 +1,323 @@
+/*-
```
Comment (plutooo) [inline nx/include/net/route.h]: Is this even exposed thru the API

File: nx/include/sys/filio.h [hunk 14]
```diff
@@ -0,0 +1,66 @@
+/*-
```
Comment (plutooo) [inline nx/include/sys/filio.h]: Will anyone ever use this?

File: nx/include/sys/_sockaddr_storage.h [hunk 15]
```diff
@@ -0,0 +1,59 @@
+/*-
```
Comment (plutooo) [inline nx/include/sys/_sockaddr_storage.h]: Will anyone ever use this?

File: nx/source/runtime/devices/socket.c [hunk 16]
```diff
@@ -0,0 +1,1550 @@
...
+    };
+
+    SfdnsresConfig sfdnsresConfig = {
+        .serialized_out_addrinfos_max_size  = config->serialized_out_addrinfos_max_size,
+        .serialized_out_hostent_max_size    = config->serialized_out_hostent_max_size,
+        .bypass_nsd                         = config->bypass_nsd,
+        .timeout                            = config->dns_timeout,
+    };
+
+    int dev = FindDevice("soc:");
```
Comment (plutooo) [inline nx/source/runtime/devices/socket.c]: What's the point of making soc: a traditional device?

File: nx/source/runtime/devices/socket.c [hunk 17]
```diff
@@ -0,0 +1,1550 @@
...
+    if(dev != -1)
+        return MAKERESULT(Module_Libnx, LibnxError_AlreadyInitialized);
+
+    ret = bsdInitialize(&bcfg);
+    if(R_SUCCEEDED(ret))
+        dev = AddDevice(&g_socketDevoptab);
+
```
Comment (plutooo) [inline nx/source/runtime/devices/socket.c]: Maybe I'm wrong but doesn't do proper cleanup on ret error here

File: nx/source/services/bsd.c [hunk 18]
```diff
@@ -81,7 +92,8 @@ static Result _bsdRegisterClient(Service* srv, TransferMemory* tmem, const BsdCo
         } *resp = r.Raw;
 
         *pid_out = resp->pid;
-        rc = resp->result;
+        g_bsdResult = rc = resp->result;
```
Comment (plutooo) [inline nx/source/services/bsd.c]: Confusing types here, typeof(errno) != Result

File: nx/source/runtime/devices/socket.c (hunk 17 above)
Comment (TuxSH) [reply nx/source/runtime/devices/socket.c]: The functions returns early a few lines above if dev != -1, thus dev == -1 in this block, so this is fine.

File: nx/include/net/bpf.h (hunk 8 above)
Comment (TuxSH) [reply nx/include/net/bpf.h]: It's supported by ioctl & needed to implement the drivers properly.

File: nx/include/net/if.h (hunk 10 above)
Comment (TuxSH) [reply nx/include/net/if.h]: Yes, by all network interface ioctls, for example it's needed to get the IP of the console (if I get around the 0xce01 issue).

File: nx/include/net/if_dl.h (hunk 11 above)
Comment (TuxSH) [reply nx/include/net/if_dl.h]: Needed to set MAC address.

File: nx/include/net/if_media.h (hunk 12 above)
Comment (TuxSH) [reply nx/include/net/if_media.h]: Needed by IFMEDIA* ioctls

File: nx/include/net/dlt.h (hunk 9 above)
Comment (TuxSH) [reply nx/include/net/dlt.h]: Dependency of bpf.h

**Pull request merged**